InvenTree/company/tests.py

```diff
@@ -174,6 +176,63 @@ def test_delete(self):
         self.assertEqual(Contact.objects.count(), 0)
 
 
+class AddressTest(TestCase):
+    """Unit tests for the Address model"""
+
+    def setUp(self):
+        """Initialization for the tests in this class"""
+        # Create a simple company
+        self.c = Company.objects.create(name='Test Corp.', description='We make stuff good')
+
+    def test_create(self):
+        """Test that object creation with only company supplied is successful"""
+        Address.objects.create(company=self.c)
+        self.assertEqual(Address.objects.count(), 1)
+
+    def test_delete(self):
+        """Test Address deletion"""
+        addr = Address.objects.create(company=self.c)
+        addr.delete()
+        self.assertEqual(Address.objects.count(), 0)
+
+    def test_primary_constraint(self):
+        """Test that there can only be one company-'primary=true' pair"""
+        c2 = Company.objects.create(name='Test Corp2.', description='We make stuff good')
+        Address.objects.create(company=self.c, primary=True)
+        Address.objects.create(company=self.c, primary=False)
+        self.assertEqual(Address.objects.count(), 2)
+
+        # Testing the constraint itself
+        # Intentionally throwing exceptions breaks unit tests unless performed in an atomic block
+        with transaction.atomic():
+            self.assertRaises(IntegrityError, Address.objects.create, company=self.c, primary=True)
```

Comment:
Comment: you could also use the following; there is a whole range of those assertations available
```suggestion
        with self.assertRaises(IntegrityError):
            Address.objects.create(company=self.c, primary=True)
```

Reply:
I tried your suggestion when working this out and it still made the test fail, as I do some DB transactions after that block.
According to my research (and testing in this case), deliberately triggering an exception within a "TestCase" method will break the method itself if you try to apply any other form of transaction within the same method afterwards. The workaround is to wrap it in the `transaction.atomic()` call as I've done here.